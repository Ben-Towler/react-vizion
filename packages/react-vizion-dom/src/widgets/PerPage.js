import { connectPerPage } from '@clinia/react-vizion-core';
import PerPage from '../components/PerPage';

/**
 * The PerPage widget displays a dropdown menu to let the user change the number
 * of displayed hits.
 *
 * If you only want to configure the number of hits per page without
 * displaying a widget, you should use the `<Configure perPage={20} />` widget. See [`<Configure />` documentation](widgets/Configure.html)
 *
 * @name PerPage
 * @kind widget
 * @propType {{value: number, label: string}[]} items - List of available options.
 * @propType {number} defaultRefinement - The number of items selected by default
 * @propType {function} [transformItems] - Function to modify the items being displayed, e.g. for filtering or sorting them. Takes an items as parameter and expects it back in return.
 * @themeKey cvi-perpage - the root div of the widget
 * @themeKey cvi-perpage-select - the select
 * @themeKey cvi-perpage-option - the select option
 * @example
 * import React from 'react';
 * import clinia from 'clinia/lite';
 * import { Vizion, PerPage, Hits } from '@clinia/react-vizion-dom';
 *
 * const searchClient = clinia(
 *   'latency',
 *   '6be0576ff61c053d5f9a3225e2a90f76'
 * );
 *
 * const App = () => (
 *   <Vizion
 *     searchClient={searchClient}
 *     indexName="instant_search"
 *   >
 *     <PerPage
 *       defaultRefinement={5}
 *       items={[
 *         { value: 5, label: 'Show 5 hits' },
 *         { value: 10, label: 'Show 10 hits' },
 *       ]}
 *     />
 *     <Hits />
 *   </Vizion>
 * );
 */

export default connectPerPage(PerPage);
