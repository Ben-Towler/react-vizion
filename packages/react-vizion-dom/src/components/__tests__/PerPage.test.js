import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import PerPage from '../PerPage';

Enzyme.configure({ adapter: new Adapter() });

describe('PerPage', () => {
  it('renders', () =>
    expect(
      renderer
        .create(
          <PerPage
            refine={() => null}
            currentRefinement={5}
            items={[
              {
                value: 5,
                label: 'show 5 hits',
              },
              {
                value: 10,
                label: 'show 10 hits',
              },
            ]}
          />
        )
        .toJSON()
    ).toMatchSnapshot());

  it('renders with a custom className', () =>
    expect(
      renderer
        .create(
          <PerPage
            className="MyCusomPerPage"
            refine={() => null}
            currentRefinement={5}
            items={[
              {
                value: 5,
                label: 'show 5 hits',
              },
              {
                value: 10,
                label: 'show 10 hits',
              },
            ]}
          />
        )
        .toJSON()
    ).toMatchSnapshot());

  it('refines its value on change', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <PerPage
        createURL={() => '#'}
        items={[
          { value: 2, label: '2 hits per page' },
          { value: 4, label: '4 hits per page' },
          { value: 6, label: '6 hits per page' },
          { value: 8, label: '8 hits per page' },
        ]}
        refine={refine}
        currentRefinement={2}
      />
    );

    const selectedValue = wrapper.find('select');
    expect(selectedValue.find('option')).toHaveLength(4);
    expect(
      selectedValue
        .find('option')
        .first()
        .text()
    ).toBe('2 hits per page');

    selectedValue.find('select').simulate('change', { target: { value: '6' } });

    expect(refine.mock.calls).toHaveLength(1);
    expect(refine.mock.calls[0][0]).toEqual('6');
  });

  it('should use value if no label provided', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <PerPage
        createURL={() => '#'}
        items={[{ value: 2 }, { value: 4 }, { value: 6 }, { value: 8 }]}
        refine={refine}
        currentRefinement={2}
      />
    );

    const selectedValue = wrapper.find('select');
    expect(selectedValue.find('option')).toHaveLength(4);
    expect(
      selectedValue
        .find('option')
        .first()
        .text()
    ).toBe('2');
  });
});
