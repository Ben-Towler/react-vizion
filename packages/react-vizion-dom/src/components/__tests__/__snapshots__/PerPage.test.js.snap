// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerPage renders 1`] = `
<div
  className="cvi-perpage"
>
  <select
    className="cvi-perpage-select"
    onChange={[Function]}
    value={5}
  >
    <option
      className="cvi-perpage-option"
      value={5}
    >
      show 5 hits
    </option>
    <option
      className="cvi-perpage-option"
      value={10}
    >
      show 10 hits
    </option>
  </select>
</div>
`;

exports[`PerPage renders with a custom className 1`] = `
<div
  className="cvi-perpage MyCusomPerPage"
>
  <select
    className="cvi-perpage-select"
    onChange={[Function]}
    value={5}
  >
    <option
      className="cvi-perpage-option"
      value={5}
    >
      show 5 hits
    </option>
    <option
      className="cvi-perpage-option"
      value={10}
    >
      show 10 hits
    </option>
  </select>
</div>
`;
